src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs

```diff
@@ -322,13 +322,33 @@ public IEnumerable<StatementSyntax> GenerateCleanupCallerAllocatedResourcesState
             if (MarshallerHelpers.GetMarshalDirection(TypeInfo, CodeContext) != MarshalDirection.ManagedToUnmanaged)
             {
                 // If we are marshalling from unmanaged to managed, we need to get the number of elements again.
+                string nativeIdentifier = context.GetIdentifiers(TypeInfo).native;
                 string numElementsIdentifier = MarshallerHelpers.GetNumElementsIdentifier(TypeInfo, context);
+
+                // Generate numElements expression with null check for pointer types
+                ExpressionSyntax numElementsExpression = ElementsMarshalling.GenerateNumElementsExpression(countInfo, countInfoRequiresCast, CodeContext, context);
+
+                // If the marshalling direction is unmanaged-to-managed and we have a native pointer type,
+                // we need to check if the native pointer is null before using the size parameter to avoid
+                // allocating arrays for null pointers.
+                if (CodeContext.Direction == MarshalDirection.UnmanagedToManaged &&
```

Comment:
I don't think we need to check the direction here as we're already under a "not ManagedToUnmanaged" check.